### ISSUE TITLE
Detail C++ coding standards

### ISSUE THREAD

**Issue by nicklafarge** [mbd-astro/forest#10]

**Comment by rjpower4** [comment]
@nicklafarge, do you think that the conventions should go in the wiki or? I'm talking about the things not enforced by clang format

**Comment by nicklafarge** [comment]
> @nicklafarge, do you think that the conventions should go in the wiki or? I'm talking about the things not enforced by clang format

I think maybe we need some sort of "contributing guide" as a part of the public-facing documentation that includes these sorts of standards definitions

**Comment by DhruvJ22** [comment]
[Drake](https://github.com/RobotLocomotion/drake) might be a good source to refer to for a contributing guide, coding style, and test setup as their codebase is primarily in C++ with a Python extension. It is highly reliable as it is maintained by TRI.

**Comment by nicklafarge** [comment]
> [Drake](https://github.com/RobotLocomotion/drake) might be a good source to refer to for a contributing guide, coding style, and test setup as their codebase is primarily in C++ with a Python extension. It is highly reliable as it is maintained by TRI.

This might be helpful for #9 too

**Comment by rjpower4** [comment]
We'll also need some sort of formatting stuff for python. Ideas? @nicklafarge, @DhruvJ22, @noahsadaka?

**Comment by DhruvJ22** [comment]
PEP 8 and for docstrings PEP 257? Check this out - [drake quide](https://drake.mit.edu/code_style_guide.html)

**Comment by noahsadaka** [comment]
[flake8](https://flake8.pycqa.org/en/latest/) for Python is a module you can run your python files through that checks for PEP 8 compliance. Could make it a requirement when submitting a pull request containing python code to run it through flake8 first and making it all PEP 8 compliant.

**Comment by nicklafarge** [comment]
Yes on PEP 8. However, the part that will complicate this is that most of the Python functionality will be written in C++, So it's a question of writing C++ code that adheres to C++ standards, but produces Python functionality that adheres to Python standards (naming contentions, documentation, etc). I found [this pybind11 style guide](https://developer.lsst.io/pybind11/style.html) from another project which could be useful for us to emulate. For documentation, I see that [pybind11 supports sphynx](https://pybind11.readthedocs.io/en/stable/advanced/misc.html#generating-documentation-using-sphinx), so that is something to consider using as well. I think flake8 is a good ideafor pull requests/our CI pipeline, but its impact will be limited to our example scripts and python unit testing, i.e. not pybind11 side of things. My guess is we are on our own with enforcing those standards there.

Thoughts?

**Comment by DhruvJ22** [comment]
I have an idea that we can use for pybind11 side of things. I will flesh it out and share it with you by tomorrow.

**Comment by rjpower4** [comment]
The auto-formatting and validation of coding standards for the example cases and testing is justification of the tooling for me. Also, I would not be surprised if there were some small shimming in python that gets added over time.

**Comment by nicklafarge** [comment]
Yeah, agreed. flake8 sounds good to me
